src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Debug/StartupProjectRegistrar.cs

```diff
@@ -0,0 +1,173 @@
+            Assumes.False(_guid == Guid.Empty);
+
+            await InitializeAsync().ConfigureAwait(false);
+        }
+
+        public async Tasks.Task InitializeAsync()
+        {
+            await AddOrRemoveProjectFromStartupProjectList(initialize: true).ConfigureAwait(false);
+
+            var watchedEvaluationRules = Empty.OrdinalIgnoreCaseStringSet.Add(ConfigurationGeneral.SchemaName);
+            var evaluationBlock = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(
+                ConfigurationGeneralRuleBlock_ChangedAsync);
+            IReceivableSourceBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>> sourceBlock = 
+                _activeConfiguredProjectSubscriptionService.ProjectRuleSource.SourceBlock;
+            _evaluationSubscriptionLink = WrapperMethodCaller.LinkTo(
+                sourceBlock,
+                evaluationBlock,
+                ruleNames: watchedEvaluationRules, suppressVersionOnlyUpdates:true
+                );
+        }
+
+        internal async Tasks.Task ConfigurationGeneralRuleBlock_ChangedAsync(IProjectVersionedValue<IProjectSubscriptionUpdate> e)
+        {
+            IProjectChangeDescription projectChange = e.Value.ProjectChanges[ConfigurationGeneral.SchemaName];
+
+            /* Currently  we watch for the change in the OutputType to check if a project is debuggable.
+               There are other cases where the OutputType will remain the same and still the ability to debuggable a project could change
+               For eg: A project's OutputType could be a Lib and an execution entry point could be added or removed
+               */
+            if (projectChange.Difference.ChangedProperties.Contains(ConfigurationGeneral.OutputTypeProperty))
+            {
+                await AddOrRemoveProjectFromStartupProjectList().ConfigureAwait(false);
+            }
+        }
+
+        private async Tasks.Task AddOrRemoveProjectFromStartupProjectList(bool initialize = false)
+        {
+            bool isDebuggable = await IsDebuggable().ConfigureAwait(false);
+
+            if (initialize || isDebuggable != _isDebuggable)
+            {
+                _isDebuggable = isDebuggable;
+                await _threadingService.SwitchToUIThread();
+                if (isDebuggable)
+                {
+                    _startupProjectsListService.AddProject(ref _guid);
+                }
+                else
+                {
+                    _startupProjectsListService.RemoveProject(ref _guid);
+                }
+            }
+        }
+
+        private async Tasks.Task<bool> IsDebuggable()
+        {
+            foreach (var provider in _launchProviders.Value.Debuggers)
+            {
+                // DebugLaunchOptions.StopAtEntryPoint seems like a valid option to use. It works.
+                // Not sure if there is a better option for this scenario.
```

Comment:
After looking into VS native code, it seems they are using 2 values either zero or `Selected|DesignTimeEvaluationBuild` and it is not clear which one is called when since the most of the call stack is optimized away. So for the time being, I am going to use `DesignTimeEvaluationBuild`, if it gives some issue then we can revisit this later.